### ISSUE TITLE
[Phase 5] CLI, Visualization, and Documentation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/predictive-maintenance-survival#6]
## Summary

Create user-friendly CLI interfaces, interactive notebooks for exploration, and complete documentation to make the system accessible to users.

**Parent Issue:** #1
**Depends on:** #5 (Phase 4: Transfer Learning)

## Objectives

- Build command-line interfaces for training and prediction
- Create exploratory Jupyter notebooks
- Document all public APIs with examples
- Provide visualization utilities

## Files to Create

| File | Description |
|------|-------------|
| `scripts/train.py` | Training CLI |
| `scripts/predict.py` | Prediction CLI |
| `scripts/evaluate.py` | Model evaluation CLI |
| `notebooks/01_data_exploration.ipynb` | Data exploration |
| `notebooks/02_model_comparison.ipynb` | Model benchmarking |
| `notebooks/03_transfer_learning.ipynb` | Transfer experiments |
| `src/visualization/plots.py` | Plotting utilities |

### CLI Scripts
- [ ] `train.py` with argparse for model selection, hyperparameters
- [ ] `predict.py` for inference on new data
- [ ] `evaluate.py` for computing metrics on test set
- [ ] Proper logging and progress bars
- [ ] Configuration file support (YAML)

### Notebooks
- [ ] Data loading and EDA
- [ ] Sensor visualization over engine lifetime
- [ ] Survival curve comparison across models
- [ ] Feature importance visualization
- [ ] Transfer learning experiments

### Visualization
- [ ] Survival curve plots with confidence bands
- [ ] Kaplan-Meier curves
- [ ] Feature importance bar charts
- [ ] RUL prediction vs actual scatter plot
- [ ] Maintenance scheduling Gantt chart

### Documentation
- [ ] Docstrings for all public functions
- [ ] Usage examples in README
- [ ] API reference (auto-generated from docstrings)

## Code Skeleton

📄 **File:** `scripts/train.py`

```python
#!/usr/bin/env python
"""Train survival models on C-MAPSS dataset."""

import argparse
import logging
from pathlib import Path
import joblib
import yaml

from src.data.loader import load_dataset, compute_rul
from src.features.extractor import SensorFeatureExtractor
from src.models.survival_forest import SurvivalForestModel
from src.models.cox_model import CoxSurvivalModel
from src.evaluation.cross_validation import CensoringAwareCV


logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Train survival model for predictive maintenance'
    )
    parser.add_argument(
        '--data-dir',
        type=Path,
        required=True,
        help='Path to C-MAPSS data directory'
    )
    parser.add_argument(
        '--subset',
        choices=['FD001', 'FD002', 'FD003', 'FD004'],
        default='FD001',
        help='Dataset subset to use'
    )
    parser.add_argument(
        '--model',
        choices=['rsf', 'cox'],
        default='rsf',
        help='Model type: rsf (Random Survival Forest) or cox (Cox PH)'
    )
    parser.add_argument(
        '--n-estimators',
        type=int,
        default=100,
        help='Number of trees for RSF'
    )
    parser.add_argument(
        '--cv-folds',
        type=int,
        default=5,
        help='Number of cross-validation folds'
    )
    parser.add_argument(
        '--output-dir',
        type=Path,
        default=Path('models'),
        help='Directory to save trained model'
    )
    parser.add_argument(
        '--config',
        type=Path,
        help='YAML config file (overrides CLI args)'
    )
    return parser.parse_args()


def main():
    args = parse_args()

    # Load config if provided
    if args.config:
        with open(args.config) as f:
            config = yaml.safe_load(f)
        # Override args with config values
        for key, value in config.items():
            setattr(args, key.replace('-', '_'), value)

    logger.info(f"Loading {args.subset} dataset from {args.data_dir}")
    train_df, test_df = load_dataset(args.data_dir, args.subset)
    train_df = compute_rul(train_df)

    logger.info("Extracting features...")
    extractor = SensorFeatureExtractor()
    X_train, y_train = extractor.fit_transform(train_df)

    # Prepare survival data
    duration = y_train['duration'].values
    event = y_train['event'].values

    logger.info(f"Training {args.model.upper()} model...")
    if args.model == 'rsf':
        model = SurvivalForestModel(n_estimators=args.n_estimators)
    else:
        model = CoxSurvivalModel()

    # Cross-validation
    logger.info(f"Running {args.cv_folds}-fold cross-validation...")
    cv = CensoringAwareCV(n_splits=args.cv_folds)
    scores = []

    for fold, (train_idx, val_idx) in enumerate(cv.split(X_train, duration, event)):
        fold_model = SurvivalForestModel(n_estimators=args.n_estimators)
        fold_model.fit(
            X_train[train_idx],
            duration[train_idx],
            event[train_idx]
        )
        score = fold_model.score(
            X_train[val_idx],
            duration[val_idx],
            event[val_idx]
        )
        scores.append(score)
        logger.info(f"  Fold {fold + 1}: C-index = {score:.4f}")

    logger.info(f"Mean C-index: {np.mean(scores):.4f} (+/- {np.std(scores):.4f})")

    # Train final model on all data
    logger.info("Training final model on all data...")
    model.fit(X_train, duration, event)

    # Save model
    args.output_dir.mkdir(parents=True, exist_ok=True)
    model_path = args.output_dir / f'{args.model}_{args.subset}.pkl'
    joblib.dump({'model': model, 'extractor': extractor}, model_path)
    logger.info(f"Model saved to {model_path}")


if __name__ == '__main__':
    main()
```

📄 **File:** `src/visualization/plots.py`

```python
"""Visualization utilities for survival analysis."""

import numpy as np
import matplotlib.pyplot as plt
from typing import Optional


def plot_survival_curve(
    surv_func,
    ax: Optional[plt.Axes] = None,
    label: str = None,
    color: str = None,
    show_ci: bool = True
):
    """Plot a single survival curve with optional confidence interval."""
    if ax is None:
        fig, ax = plt.subplots(figsize=(10, 6))

    times = surv_func.x
    probs = surv_func.y

    ax.step(times, probs, where='post', label=label, color=color)
    ax.set_xlabel('Time (cycles)')
    ax.set_ylabel('Survival Probability')
    ax.set_ylim(0, 1.05)
    ax.legend()

    return ax


def plot_rul_comparison(
    actual_rul: np.ndarray,
    predicted_rul: np.ndarray,
    lower_bound: Optional[np.ndarray] = None,
    upper_bound: Optional[np.ndarray] = None,
    ax: Optional[plt.Axes] = None
):
    """Scatter plot of predicted vs actual RUL."""
    if ax is None:
        fig, ax = plt.subplots(figsize=(8, 8))

    ax.scatter(actual_rul, predicted_rul, alpha=0.6, s=30)

    if lower_bound is not None and upper_bound is not None:
        for i in range(len(actual_rul)):
            ax.plot(
                [actual_rul[i], actual_rul[i]],
                [lower_bound[i], upper_bound[i]],
                color='gray', alpha=0.3
            )

    # Perfect prediction line
    max_val = max(actual_rul.max(), predicted_rul.max())
    ax.plot([0, max_val], [0, max_val], 'r--', label='Perfect prediction')

    ax.set_xlabel('Actual RUL')
    ax.set_ylabel('Predicted RUL')
    ax.legend()

    return ax


def plot_feature_importance(
    feature_names: list,
    importance_values: np.ndarray,
    top_k: int = 20,
    ax: Optional[plt.Axes] = None
):
    """Horizontal bar chart of feature importance."""
    if ax is None:
        fig, ax = plt.subplots(figsize=(10, 8))

    # Sort and take top k
    idx = np.argsort(importance_values)[-top_k:]
    names = [feature_names[i] for i in idx]
    values = importance_values[idx]

    ax.barh(names, values)
    ax.set_xlabel('Importance')
    ax.set_title(f'Top {top_k} Features')

    return ax
```

## Definition of Done

- [ ] `python scripts/train.py --help` shows all options
- [ ] Training completes without errors on all subsets
- [ ] Notebooks run end-to-end without errors
- [ ] All public functions have docstrings with examples
- [ ] Visualizations render correctly

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Notebook kernel dies on large data | 🟢 | Use subset for exploration |
| Argparse complexity | 🟢 | Use config files for complex setups |
| Missing edge cases in CLI | 🟡 | Comprehensive testing |
